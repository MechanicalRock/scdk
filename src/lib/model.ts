// types for rendering the constructs
export type Product = {
    Name: string;
    ProvisioningArtifacts: ProvisioningArtifact[];
};

export type ProvisioningArtifact = {
    Name: string;
    Parameters: ProvisioningArtifactParameter[];
    Outputs: {
        OutputName: string;
        OutputDescription?: string;
    }[];
};

export type ProvisioningArtifactParameter = {
    ParameterName: string;
    ParameterType: string;
    ParameterRequired: boolean;
    ParameterDescription?: string;
};

export type Portfolio = {
    Products: Product[];
};

// Types for parsing CloudFormation templates
export type Template = {
    Parameters: { [key: string]: Parameter };
    Outputs: { [key: string]: Output };
};

export type Parameter = {
    Type: ParameterType;
    Description?: string;
    Default?: string;
};

export type ParameterType = string;

export type Output = {
    Description?: string;
};
