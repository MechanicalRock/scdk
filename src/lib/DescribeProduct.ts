import { Observable, defer, EMPTY } from 'rxjs';
import { map, flatMap, expand } from 'rxjs/operators'
import { ServiceCatalog } from 'aws-sdk'

export function SearchProducts$(catalog: ServiceCatalog, portfolioId: string, product: string): Observable<ServiceCatalog.ProductViewDetail> {
    const filter = !product ? null : {
        'FullTextSearch': [product]
    }

    const searchProducts$ = (page?: string): Observable<ServiceCatalog.SearchProductsAsAdminOutput> =>
        defer(() => catalog.searchProductsAsAdmin({ PortfolioId: portfolioId, PageToken: page, Filters: filter }).promise())

    return searchProducts$()
        .pipe(expand<ServiceCatalog.SearchProductsAsAdminOutput>(({ NextPageToken }) => NextPageToken ? 
            searchProducts$(NextPageToken): EMPTY))
        .pipe(flatMap(r => r.ProductViewDetails))
}

export function DescribeProductAsAdmin$(catalog: ServiceCatalog, productId: string): Observable<ServiceCatalog.DescribeProductAsAdminOutput> {
    return defer(() => catalog.describeProductAsAdmin({Id: productId}).promise())
}

export function DescribeProvisioningParameters$(catalog: ServiceCatalog, productId: string, artifactId: string): Observable<ServiceCatalog.ProvisioningArtifactParameters> {
    const describeParameters$ = (): Observable<ServiceCatalog.DescribeProvisioningParametersOutput> => 
        defer(() => catalog.describeProvisioningParameters({ 
            ProductId: productId, ProvisioningArtifactId: artifactId }).promise())
    
    return describeParameters$().pipe(map(r => r.ProvisioningArtifactParameters))
}
