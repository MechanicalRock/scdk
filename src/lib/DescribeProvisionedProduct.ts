import { Observable, defer, EMPTY } from 'rxjs';
import { flatMap, map, expand } from 'rxjs/operators'
import { ServiceCatalog, CloudFormation } from 'aws-sdk'

export function SearchProvisionedProducts$(catalog: ServiceCatalog, product: string): Observable<ServiceCatalog.ProvisionedProductAttribute> {
    const filter = !product ? null : {
        'SearchQuery': [`name:${product}`]
    }

    const searchProvisionedProducts$ = (page?: string): Observable<ServiceCatalog.SearchProvisionedProductsOutput> =>
        defer(() => catalog.searchProvisionedProducts({ PageToken: page, Filters: filter }).promise())

    return searchProvisionedProducts$()
        .pipe(expand<ServiceCatalog.SearchProvisionedProductsOutput>(({ NextPageToken }) => NextPageToken ? 
            searchProvisionedProducts$(NextPageToken): EMPTY))
        .pipe(flatMap(r => r.ProvisionedProducts))
}

export function DescribeRecordOutputs$(catalog: ServiceCatalog, recordId: string): Observable<ServiceCatalog.RecordOutputs> {
    return defer(() => catalog.describeRecord({ Id: recordId}).promise())
        .pipe(
            map(r => r.RecordOutputs)
        )
}

export function DescribeProvisioningArtifact$(catalog: ServiceCatalog, productId: string, artifactId: string): Observable<ServiceCatalog.ProvisioningArtifactDetail> {
    return defer(() => catalog.describeProvisioningArtifact({ ProvisioningArtifactId: artifactId, ProductId: productId}).promise())
        .pipe(map(r => r.ProvisioningArtifactDetail))
}

export function DescribeParameters$(cfn: CloudFormation, stackArn: string) {
    return defer(() => cfn.describeStacks({ StackName: stackArn}).promise())
        .pipe(
            map(r => r.Stacks[0].Parameters)
        )
}