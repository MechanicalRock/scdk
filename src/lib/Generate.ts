
// nunjucks.configure('.', {})

// const source = nunjucks.render('something.ts.njk', test) 

interface ProductAst {
    name: string
    productId: string
    version: string
    parameters: ProductParameter[]
    outputs: ProductOutput[]
}

interface ProductParameter {
    name: string
    default?: string
}

interface ProductOutput {
    name: string
}

interface Renderer {
    render(template: string, product: ProductAst): string
}

export class CdkProductRenderer {
    constructor(private renderer: Renderer) {}

    generate(template: string, product: ProductAst): string {
        return this.renderer.render(template, product)
    }
}