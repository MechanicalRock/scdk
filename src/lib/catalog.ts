import ServiceCatalog from "aws-sdk/clients/servicecatalog";
import S3 from "aws-sdk/clients/s3";
import ok from "assert";
import YAML from "js-yaml";
import { Product, Template, ProvisioningArtifactParameter, ProvisioningArtifact } from "./model";

export async function all<T>(generator: AsyncGenerator<T, void, unknown>): Promise<T[]> {
    const items: T[] = [];
    for await (const item of generator) {
        items.push(item);
    }
    return items;
}

export class Catalog {
    constructor(private catalog: ServiceCatalog, private s3: S3) {}

    async *listProducts(PortfolioId?: string): AsyncGenerator<Product, void, unknown> {
        let PageToken;
        do {
            const { NextPageToken, ProductViewDetails } = await this.catalog
                .searchProductsAsAdmin({ PortfolioId })
                .promise();
            ok(ProductViewDetails);
            for (const { ProductViewSummary } of ProductViewDetails) {
                ok(ProductViewSummary?.Name);
                ok(ProductViewSummary?.ProductId);
                const { Name, ProductId } = ProductViewSummary;
                const ProvisioningArtifacts = await all(this.listProvisioningArtifacts(ProductId));
                yield { Name, ProvisioningArtifacts };
            }
            PageToken = NextPageToken;
        } while (PageToken);
    }

    async *getProducts(ProductIds: string[]): AsyncGenerator<Product, void, unknown> {
        for (const Id of ProductIds) {
            const { ProductViewSummary } = await this.catalog.describeProduct({ Id }).promise();
            ok(ProductViewSummary?.Name);
            ok(ProductViewSummary?.ProductId);
            const { Name, ProductId } = ProductViewSummary;
            const ProvisioningArtifacts = await all(this.listProvisioningArtifacts(ProductId));
            ok(ProvisioningArtifacts);
            yield { Name: Name, ProvisioningArtifacts };
        }
    }

    private async *listProvisioningArtifacts(ProductId: string): AsyncGenerator<ProvisioningArtifact, void, unknown> {
        const { ProvisioningArtifacts } = await this.catalog.describeProduct({ Id: ProductId }).promise();
        ok(ProvisioningArtifacts);
        for (const { Id: ProvisioningArtifactId, Name } of ProvisioningArtifacts) {
            ok(Name);
            ok(ProvisioningArtifactId);
            const { Info } = await this.catalog
                .describeProvisioningArtifact({ ProductId, ProvisioningArtifactId })
                .promise();
            ok(Info?.TemplateUrl);
            const { Parameters, Outputs } = await this.loadTemplateFromURL(Info.TemplateUrl);
            yield { Name, Parameters, Outputs };
        }
    }

    private async loadTemplateFromURL(url: string) {
        const re = /^https:\/\/(?<Bucket>.+)\.s3([\.-]?(?<Region>.+))?\.amazonaws\.com\/(?<Key>.+(json|yaml|yml))$/;
        const match = re.exec(url);
        const groups = match?.groups;
        ok(groups?.Bucket);
        ok(groups?.Key);
        const { Bucket, Key } = groups;
        this.s3.config.region = groups?.Region ?? this.s3.config.region;
        const { Body } = await this.s3.getObject({ Bucket, Key }).promise();
        ok(Body);
        const templateObject: Partial<Template> = Key.endsWith("json")
            ? JSON.parse(Body.toString())
            : YAML.safeLoad(Body.toString());
        return {
            Parameters: this.getParameters(templateObject),
            Outputs: this.getOutputs(templateObject),
        };
    }

    private getParameters(template: Partial<Template>): ProvisioningArtifactParameter[] {
        if (!template?.Parameters) {
            return [];
        }
        return Object.entries(template.Parameters).map(([ParameterName, Parameter]) => ({
            ParameterName,
            ParameterType: Parameter.Type,
            ParameterRequired: Parameter.Default === undefined,
            ParameterDescription: Parameter.Description,
        }));
    }

    private getOutputs(template: Partial<Template>): { OutputName: string }[] {
        if (!template?.Outputs) {
            return [];
        }
        return Object.entries(template.Outputs).map(([OutputName, Output]) => ({
            OutputName,
            OutputDescription: Output.Description,
        }));
    }
}
