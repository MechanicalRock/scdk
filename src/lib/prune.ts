import { Portfolio, ProvisioningArtifact } from "./model";
import semver, { major } from "semver";

const RE = /^v(?<major>\d)+(\.(?<minor>\d+))(\.(?<patch>\d+))$/;

function sort(artifacts: ProvisioningArtifact[]): ProvisioningArtifact[] {
    const tmp = artifacts.filter((pa) => RE.test(pa.Name)).sort((a, b) => (semver.gt(a.Name, b.Name) ? -1 : 1));
    const majors = Array.from(new Set(tmp.map((pa) => major(pa.Name))));
    return majors.map((major) => tmp.find((pa) => pa.Name.startsWith(`v${major}`))) as any;
}

export function pruneVersions({ Products }: Portfolio): Portfolio {
    const products = Products.map(({ Name, ProvisioningArtifacts }) => ({
        Name,
        ProvisioningArtifacts: sort(ProvisioningArtifacts),
    }));
    return {
        Products: products.filter(({ ProvisioningArtifacts: pa }) => pa.length > 0),
    };
}
