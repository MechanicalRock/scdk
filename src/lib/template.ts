import { Environment, FileSystemLoader } from "nunjucks";
import pascalcase from "pascalcase";
import { join } from "path";
import { promises as fs, existsSync, mkdirSync } from "fs";
import { Product, Portfolio } from "./model";
import { major } from "semver";

const njs = new Environment(new FileSystemLoader(join(__dirname, "../..", "templates")));

export async function writeTemplates(portfolio: Portfolio, baseDir: string): Promise<void[]> {
    if (!existsSync(baseDir)) {
        mkdirSync(baseDir);
    }
    return await Promise.all(
        Object.entries(getTemplates(portfolio)).map(([filename, template]) =>
            fs.writeFile(join(baseDir, filename), template),
        ),
    );
}

export function getTemplates(portfolio: Portfolio): { [key: string]: string } {
    const index = {
        ["index.ts"]: renderPortfolio(portfolio),
    };
    const fileContents = portfolio.Products.map((p) => ({
        [`${pascalcase(p.Name)}.ts`]: renderProduct(p),
    }));
    return Object.assign(index, ...fileContents);
}

export const renderProduct = (product: Product): string => {
    return njs.render("product.ts.njk", { ...product, pascalcase, major });
};

export const renderPortfolio = (portfolio: Portfolio): string => {
    return njs.render("portfolio.ts.njk", { ...portfolio, pascalcase });
};
