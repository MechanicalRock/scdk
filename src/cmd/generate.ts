import { Argv, Arguments } from 'yargs'
import { ServiceCatalog } from 'aws-sdk'
import { ProductAst } from '../lib/Generate'
import compareVersions from 'compare-versions';
import { pascalCase } from 'change-case'

import { CdkProductRenderer } from '../lib/generate';
import { join } from 'path';
import * as nunjucks from 'nunjucks';

type Args = {
    portfolio: Promise<string>
    response: any
}

module.exports = {
    command: 'generate',
    describe: 'generate CDK constructs for the target catalog',
    builder: (yargs: Argv) => {
        return yargs.demandOption('portfolio')
            .coerce('portfolio', async (arg: string) => {
                if (arg.startsWith('port-')) {
                    return arg
                }
                const catalog = new ServiceCatalog()
                let nextPage = () => ListPortfolios(catalog)
                do {
                    const { PortfolioDetails, nextPage: next }  = await nextPage()
                    nextPage = next

                    const portfolio = PortfolioDetails.find(pd => pd.DisplayName === arg)
                    if (portfolio) {
                        return portfolio.Id
                    }
                } while (nextPage)
                throw new Error('Could not find specified portfolio')
            })
    },
    handler: async (args: Arguments<Args>) => {
        const portfolio = await args.portfolio
        const catalog = new ServiceCatalog()
        const asts: ProductAst[] = []
        let nextPage = () => SearchProducts(catalog, portfolio)

        do {
            const { ProductViewDetails: pvds, nextPage: next } = await nextPage()

            const productDetails = (await Promise.all(pvds.map(async p => {
                const {ProductViewDetail, ProvisioningArtifactSummaries} = 
                    await ListProductDetails(catalog, p.ProductViewSummary.ProductId)
                
                return ProvisioningArtifactSummaries.map(ProvisioningArtifactSummary => ({
                    ProvisioningArtifactSummary,
                    ProductViewDetail
                }))
                .sort((a, b) => versionSort(b.ProvisioningArtifactSummary.Name, a.ProvisioningArtifactSummary.Name))
                .slice(0, 1)
            }))).reduce((acc, n) => acc.concat(n))

            const data = await Promise.all(productDetails.map(async p => {
                const ProvisioningParameters = await GetProvisioningParameters(
                    catalog, p.ProductViewDetail.ProductViewSummary.ProductId, p.ProvisioningArtifactSummary.Id)

                return {
                    ...p,
                    ProvisioningParameters
                }
            }))
            
            asts.push(...data.map(d => convertToAst(d)))
            nextPage = next
        } while (nextPage)

        const templateDir = join(__dirname, '../../templates')
        const renderer = nunjucks.configure(templateDir, {})
        const cdkProductRenderer = new CdkProductRenderer(renderer)
        const source = cdkProductRenderer.generate('cdk-product.ts.njk', { products: asts })
        args.response(source)
    }
}

interface AwsModel {
    ProvisioningParameters: ServiceCatalog.ProvisioningArtifactParameter[];
    ProductViewDetail: ServiceCatalog.ProductViewDetail;
    ProvisioningArtifactSummary: ServiceCatalog.ProvisioningArtifactSummary;
}

export function convertToAst(model: AwsModel): ProductAst {
    return {
        name: pascalCase(model.ProductViewDetail.ProductViewSummary.Name),
        productId: model.ProductViewDetail.ProductViewSummary.ProductId,
        version: model.ProvisioningArtifactSummary.Name,
        parameters: model.ProvisioningParameters.map(p => ({ name: p.ParameterKey, default: p.DefaultValue})),
    }
}

function versionSort(va: string, vb: string) {
    const a: string = (va.match(/v[\d\.?]+/) as any)[0];
    const b: string = (vb.match(/v[\d\.?]+/) as any)[0];
    return compareVersions(a, b);
}

export async function SearchProducts(catalog: ServiceCatalog, portfolioId: string) {
    // const filter = !product ? null : {
    //     'FullTextSearch': [product]
    // }

    const searchProducts = async (page?: string) => {
        const { ProductViewDetails, NextPageToken} = await catalog.searchProductsAsAdmin({ PortfolioId: portfolioId, PageToken: page}).promise()
        const nextPage = NextPageToken ? () => searchProducts(NextPageToken) : undefined
        return {
            ProductViewDetails,
            nextPage
        }
    }
    return searchProducts()
}

export async function ListPortfolios(catalog: ServiceCatalog) {
    const listPortfolios = async (page?: string) => {
        const { PortfolioDetails, NextPageToken} = await catalog.listPortfolios({ PageToken: page }).promise()
        const nextPage = NextPageToken ? () => listPortfolios(NextPageToken) : undefined
        console.log(PortfolioDetails)
        return {
            PortfolioDetails,
            nextPage
        }
    }
    return listPortfolios()
}

export async function ListProductDetails(catalog: ServiceCatalog, productId: string) {
    const { ProductViewDetail, ProvisioningArtifactSummaries} = await catalog.describeProductAsAdmin({Id: productId}).promise()
    return { ProductViewDetail, ProvisioningArtifactSummaries}
}

export async function GetProvisioningParameters(catalog: ServiceCatalog, productId: string, artifactId: string) {
    const PathId = (await catalog.listLaunchPaths({ ProductId: productId }).promise()).LaunchPathSummaries[0].Id
    const result = await catalog.describeProvisioningParameters({ 
        ProductId: productId, ProvisioningArtifactId: artifactId, PathId }).promise()
    return result.ProvisioningArtifactParameters
}
