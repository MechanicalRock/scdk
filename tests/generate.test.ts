import { CdkProductRenderer } from '../src/lib/generate';
import { join } from 'path';
import * as nunjucks from 'nunjucks';

describe('Given an AST representing the service catalog product', () => {
    const input = [
        {
            name: 'MyProduct',
            version: 'v1',
            productId: 'prod-xyz',
            parameters: [
                {
                    name: 'parameter_one',
                    default: 'one'
                },
                {
                    name: 'parameter_two'
                }
            ],
            outputs: [
                {
                    name: 'output_one'
                },
                {
                    name: 'output_two'
                }
            ]
        }
    ]

    describe('When I run generate against the AST', () => {
        const templateDir = join(__dirname, '../src/lib/templates')
        console.log(templateDir)
        const renderer = nunjucks.configure(templateDir, {})
        const cdkProductRenderer = new CdkProductRenderer(renderer)
        const source = cdkProductRenderer.generate('cdk-product.ts.njk', input[0])

        it('then it should produce the correct typescript source', () => {
            expect(source).toMatchSnapshot()
        })
    })
})