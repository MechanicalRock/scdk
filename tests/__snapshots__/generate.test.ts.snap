// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an AST representing the service catalog product When I run generate against the AST then it should produce the correct typescript source 1`] = `
"import { CfnCloudFormationProvisionedProduct } from '@aws-cdk/aws-servicecatalog';
import { Construct, Fn } from '@aws-cdk/cdk';

export namespace MyProduct {
    interface v1Props {
        parameter_one?: string
        parameter_two: string
    }

    type v1Outputs = 'output_one' | 'output_two';

    export class v1 extends Construct {
        private _provisionedProduct: CfnCloudFormationProvisionedProduct

        /** Output returns an appropriate exported stack value for the given key **/
        public Output: { readonly [key in v1Outputs ]: string}

        /** Create a v1 MyProduct component */
        constructor(scope: Construct, id: string, props?: v1Props) {
            super(scope, id)
    
            const provisioningParameters = !props ? 
                [] : Object.keys(props).map(key => ({[key]: key, value: props[key]}))
    
            this._provisionedProduct = new CfnCloudFormationProvisionedProduct(scope, id, {
                productId: 'prod-xyz',
                provisioningArtifactName: 'v1',
                provisioningParameters
            })
            this.Output = {
                output_one: this.Export('output_one')
                ,output_two: this.Export('output_two')
            }
        }
    
        public Export(value: v1Outputs) {
            const attr = this._provisionedProduct.getAtt('CloudformationStackArn')
            return Fn.importValue(Fn.sub('\${provisionedProductStackArn}/\${exportedValue}', {
                'provisionedProductStackArn' : attr.toString(),
                'exportedValue': value
            }))
        }
    }
}
"
`;
